File: app/src/main/java/com/challenge/fidoreader/fagment/ChangePINFragment.java
package com.challenge.fidoreader.fagment;

import android.app.AlertDialog;
import android.app.Dialog;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.fragment.app.DialogFragment;

import com.challenge.fidoreader.MainActivity;
import com.challenge.fidoreader.R;
import com.challenge.fidoreader.Util.Util;


public class ChangePINFragment extends DialogFragment {

    public String originPIN = "";
    public String clientPIN = "";
    private MainActivity activity;

    private View view;
    private EditText originPINText;
    private EditText password1;
    private EditText password2;

    private TextView resultView;

    private Button okbtn;
    private ImageView cancelbtn;

    public ChangePINFragment(MainActivity activity){
        this.activity = activity;
    }

    public Dialog onCreateDialog(Bundle savedInstanceState){

        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        LayoutInflater inflater = getActivity().getLayoutInflater();
        view = inflater.inflate(R.layout.changepin_popup, null);

        builder.setView(view);

        return builder.create();

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        // Dialog Title 없애기
        getDialog().requestWindowFeature(Window.FEATURE_NO_TITLE);
        view = inflater.inflate(R.layout.changepin_popup, null);
        getDialog().setContentView(view);

        originPINText = (EditText)view.findViewById(R.id.beforePINEditText);
        password1 = (EditText)view.findViewById(R.id.newPINEditText);
        password2 = (EditText)view.findViewById(R.id.confrimPINEditText);

        resultView = (TextView)view.findViewById(R.id.resultSetView);

        password2.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if(password1.getText().toString().equals(password2.getText().toString())){
                    resultView.setTextColor(Color.parseColor("#00FF00"));
                    resultView.setText("Matched Password");
                }
                else{
                    resultView.setTextColor(Color.parseColor("#FF0000"));
                    resultView.setText("Not Matched Password");
                }
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        okbtn = (Button)view.findViewById(R.id.OKBtn);
        okbtn.setOnClickListener(new View.OnClickListener(){

            @Override
            public void onClick(View v) {
                if(originPINText.getText().toString().length() > 0) {
                    try {
                        originPIN = Util.ascii(originPINText.getText().toString());
                        clientPIN = Util.ascii(password1.getText().toString());
                        String result = activity.authenticator.changePin(originPIN, clientPIN);

                        if(result.equals("00")) {
                            Toast.makeText(activity.getApplicationContext(), "PIN 변경이 완료되었습니다.", Toast.LENGTH_LONG).show();

                            getDialog().dismiss();
                        }
                        else if(result.equals("31")){
                            Toast.makeText(activity.getApplicationContext(), "기존 PIN 정보가 올바르지 않습니다.", Toast.LENGTH_LONG).show();
                        }
                        else{
                            Toast.makeText(activity.getApplicationContext(), "PIN 변경을 정상적으로 종료하지 못하였습니다.", Toast.LENGTH_LONG).show();

                            getDialog().dismiss();
                        }

                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
            }
        });

        cancelbtn = (ImageView) view.findViewById(R.id.CancelBtn);
        cancelbtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                try{
                    Toast.makeText(activity.getApplicationContext(), "PIN 변경을 취소하였습니다.", Toast.LENGTH_LONG).show();
                    getDialog().dismiss();

                } catch (Exception e){
                    e.printStackTrace();;
                }
            }
        });

        // 이미지랑 Layout이랑 안겹치는 부분 배경색 없애기
        getDialog().getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        getDialog().show();

        return view;
    }

}

File: app/src/main/java/com/challenge/fidoreader/fagment/ChangePINFragment.kt
package com.challenge.fidoreader.fagment

import android.app.AlertDialog
import android.app.Dialog
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.widget.*
import androidx.fragment.app.DialogFragment
import com.challenge.fidoreader.MainActivity
import com.challenge.fidoreader.R
import com.challenge.fidoreader.Util.ascii
import com.challenge.fidoreader.fidoReader.CBOR_CODE_CTAP1_ERR_SUCCESS
import com.challenge.fidoreader.fidoReader.CBOR_CODE_CTAP2_ERR_PIN_INVALID


class ChangePINFragment(val activity: MainActivity) : DialogFragment() {

/*
    lateinit var mainView: View
    lateinit var originPINText: EditText
    lateinit var password1: EditText
    lateinit var password2: EditText

    lateinit var resultView: TextView

    lateinit var okbtn: Button
    lateinit var cancelbtn: ImageView


    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val builder = AlertDialog.Builder(getActivity())
        val inflater = getActivity()!!.layoutInflater
        mainView = inflater.inflate(R.layout.changepin_popup, null)

        builder.setView(mainView)
        return builder.create()
    }
*/

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        var mainView: View = inflater.inflate(R.layout.changepin_popup, null)

        // Dialog Title 없애기
        dialog!!.requestWindowFeature(Window.FEATURE_NO_TITLE)
        dialog!!.setContentView(mainView)

        var originPINText = mainView.findViewById<EditText>(R.id.beforePINEditText)
        var password1 = mainView.findViewById<EditText>(R.id.newPINEditText)
        var password2 = mainView.findViewById<EditText>(R.id.confrimPINEditText)

        var resultView = mainView.findViewById<TextView>(R.id.resultSetView)

        password2.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable) {}
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                if (password1.text.toString() == password2.text.toString()) {
                    resultView.setTextColor(Color.parseColor("#00FF00"))
                    resultView.text = "Matched Password"
                } else {
                    resultView.setTextColor(Color.parseColor("#FF0000"))
                    resultView.text = "Not Matched Password"
                }
            }
        })

        var originPIN = ""
        var clientPIN = ""

        var okbtn = mainView.findViewById<Button>(R.id.OKBtn)
        okbtn.setOnClickListener {
            if (originPINText.text.toString().isNotEmpty()) {
                try {
                    originPIN = originPINText.text.toString().ascii()
                    clientPIN = password1.text.toString().ascii()
                    val result = MainActivity.authenticator.changePin(originPIN, clientPIN)
                    when(result){
                        CBOR_CODE_CTAP1_ERR_SUCCESS ->{
                            Toast.makeText(activity.applicationContext, "PIN 변경이 완료되었습니다.", Toast.LENGTH_LONG).show()
                            dialog!!.dismiss()
                        }
                        CBOR_CODE_CTAP2_ERR_PIN_INVALID ->{
                            Toast.makeText(activity.applicationContext, "기존 PIN 정보가 올바르지 않습니다.", Toast.LENGTH_LONG).show()
                        }
                        else ->{
                            Toast.makeText(activity.applicationContext, "PIN 변경을 정상적으로 종료하지 못하였습니다.", Toast.LENGTH_LONG).show()
                            dialog!!.dismiss()
                        }
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }

        var cancelbtn = mainView.findViewById<ImageView>(R.id.CancelBtn)
        cancelbtn.setOnClickListener {
            try {
                Toast.makeText(activity.applicationContext, "PIN 변경을 취소하였습니다.", Toast.LENGTH_LONG).show()
                dialog!!.dismiss()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        // 이미지랑 Layout이랑 안겹치는 부분 배경색 없애기
        dialog!!.window!!.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        dialog!!.show()

        return mainView
    }

}
